### ISSUE TITLE
Tests failing after coroutines update

### ISSUE THREAD

**Issue by Alexs784** [mockk/mockk#1094]
### Prerequisites

- [x] I am running the latest version
- [x] I checked the documentation and found no answer
- [x] I checked to make sure that this issue has not already been filed

### Expected Behavior

The tests should pass and irrelevant Mockk exceptions should be ignored as it used to be (?)

### Current Behavior

Multiple tests failing

### Failure Information (for bugs)

The errors I am seeing in the console are a `UncaughtExceptionsBeforeTest` followed by a 
```
Suppressed: io.mockk.MockKException: no answer found for MyMockedClass
[...]
Caused by: io.mockk.MockKException: no answer found for MyMockedClass
```

#### Steps to Reproduce

1. Have a project with some tests that use Mockk
2. Have coroutines 1.6.4 and verify tests are passing
3. Update coroutines to 1.7.1 and verify tests are failing

Note: what I also noticed is that usually is only one test failing, and if I try to delete that one just to make sure that's the problem, another one fails. And so on. I am pretty sure this is related to [this coroutines change of behaviour in the new version](https://github.com/Kotlin/kotlinx.coroutines/issues/1205), and this one as well https://github.com/Kotlin/kotlinx.coroutines/issues/3738

**Note 2: this happens only when running all the tests together, or if multiple tests are run within a single file. This never happens if the tests are run singularly**

#### Context

* MockK version: 1.13.5
* OS: Mac Ventura 13.4 - Android Studio
* Kotlin version: 1.8.21
* JDK version: 17
* JUnit version: jUnit5 - 5.9.2
* Type of test: unit test

#### Stack trace
```
Suppressed: io.mockk.MockKException: no answer found for MyMockedClass
[...]
Caused by: io.mockk.MockKException: no answer found for MyMockedClass
```

**Comment by albinus-sc** [comment]
Any progress on this? Or any ideas for workarounds?

**Comment by kurteous** [comment]
We're seeing the same thing: multiple failures with previously-working tests after updating to Kotlin 1.8.22, which updates coroutines from 1.6.4 to 1.7.1. The failures are `kotlinx.coroutines.test.UncaughtExceptionsBeforeTest` followed by `Suppressed: io.mockk.MockKException: no answer found for: ...`, as noted above.

**Comment by atomgomba** [comment]
We're also affected by this. Running a single test works, but running multiple tests (say each test in a class) throws `UncaughtExceptionsBeforeTest`.

**Comment by vvictor10** [comment]
Seeing the same issue. Running single test class works, but running a suite or running all of it in a module on CI breaks it.

**Comment by jczerski** [comment]
I'm facing with same issue 😢

**Comment by sunragav** [comment]
What is the solution for this?

**Comment by kurteous** [comment]
In our case, the solution was to fix all the issues with our unit tests. These errors existed before updating coroutines, but they didn't cause the unit tests to fail. Rather, the errors were output to the console and were not visible unless you scrolled to the bottom of a lot of log lines. Because the tests passed we didn't know there were any issues.

After updating coroutines the tests failed, at least when run in groups, so we were forced to address all the issues. This is a good thing. Most of the issues were the `No answer for...` types described above, though we did find other issues, and we discovered that some unit tests were not correct once we provided all mocks so the tests could complete without throwing Exceptions in coroutines.

Though it was a pain to fix the issues, our code is in much better shape now and we have a lot more confidence in our tests.

**Comment by atomgomba** [comment]
> In our case, the solution was to fix all the issues with our unit tests. These errors existed before updating coroutines, but they didn't cause the unit tests to fail. Rather, the errors were output to the console and were not visible unless you scrolled to the bottom of a lot of log lines. Because the tests passed we didn't know there were any issues.
> 
> After updating coroutines the tests failed, at least when run in groups, so we were forced to address all the issues. This is a good thing. Most of the issues were the `No answer for...` types described above, though we did find other issues, and we discovered that some unit tests were not correct once we provided all mocks so the tests could complete without throwing Exceptions in coroutines.
> 
> Though it was a pain to fix the issues, our code is in much better shape now and we have a lot more confidence in our tests.

Interesting! How does this explain the fact that the tests are successful when run individually?

**Comment by atomgomba** [comment]
> I don't know why that happens, but I suspect if you scroll to the bottom of the console output you will see issues whether the tests are run singly or in a group. And once you fix the issues, the tests will pass when run in a group and when run singly.

Thank you for sharing this! So if this is the case I think adding this to the build file might help finding the culprit:

```kotlin
tasks.withType<Test> {
    testLogging {
        exceptionFormat = TestExceptionFormat.FULL
        showStackTraces = true
    }
}
```

**Comment by atomgomba** [comment]
> > > I don't know why that happens, but I suspect if you scroll to the bottom of the console output you will see issues whether the tests are run singly or in a group. And once you fix the issues, the tests will pass when run in a group and when run singly.
> > 
> > 
> > Thank you for sharing this! So if this is the case I think adding this to the build file might help finding the culprit:
> > ```kotlin
> > tasks.withType<Test> {
> >     testLogging {
> >         exceptionFormat = TestExceptionFormat.FULL
> >         showStackTraces = true
> >     }
> > }
> > ```
> 
> Great idea! Looks like `showStandardStreams` might help, too. It logs stdOut and stdErr as well. See https://docs.gradle.org/current/dsl/org.gradle.api.tasks.testing.logging.TestLogging.html for more options.

By the way it is more convenient to observe stack traces in the generated HTML report than in the terminal. It seems like they can get truncated in the terminal.

EDIT: So it looks like `runTest` is checking for previous exceptions [before executing](https://github.com/Kotlin/kotlinx.coroutines/blob/5d4954b17ac18c572e22a9a6e7bb4e4368ec568c/kotlinx-coroutines-test/common/src/TestScope.kt#L236) the provided block and this could explain why a test won't fail when run individually, but it can indeed fail when multiple tests are run and at least one of them threw a suppressed exception. Furthermore this suggest that the same `TestScope` is used for a single test run, regardless of the number of tests involved.

**Comment by atomgomba** [comment]
I can also confirm that failures were due to previously suppressed exceptions in the tests. Adding this to the build file helped a lot in identifying and fixing them:
```kotlin
tasks.withType<Test> {
    testLogging {
        exceptionFormat = TestExceptionFormat.FULL
        showStackTraces = true
        showStandardStreams = true
    }
}
```
In our case the failing tests were mostly the ones that were using the Roboelectric runner.

**Comment by kurteous** [comment]
> EDIT: So it looks like `runTest` is checking for previous exceptions [before executing](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-test/common/src/TestScope.kt#L242) the provided block and this could explain why a test won't fail when run individually, but it can indeed fail when multiple tests are run and at least one of them threw a suppressed exception. Furthermore this suggest that the same `TestScope` is used for a single test run, regardless of the number of tests involved.

Great discovery! Mystery solved.

**Comment by grassehh** [comment]
What is the exceptions are thrown by a third party library, cannot be fixed and are not blocking ? 
How are we supposed to fix this please ?

**Comment by hamid97m** [comment]
Is there any other way? fixing all issues is not possible :((

**Comment by grassehh** [comment]
> Is there any other way? fixing all issues is not possible :((

For now, the workaround we use is to use "[runBlocking](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/run-blocking.html)" instead of "runTest".

**Comment by pitoszud** [comment]
In my case I didn't include coEvery{} for some repository calls, because I wasn't getting an error such as "no answer found for MyRepo(MyRepo#1).myfunction(), but after adding coEvery { myRepo.myFunction() }... both individual and a multiple tests have passed (before the change only a single test were passing).
So, my recommendation is to double check, you have your coEvery or every included in all tests even if they pass fine without it.

**Comment by siddhi-droid** [comment]
> I can also confirm that failures were due to previously suppressed exceptions in the tests. Adding this to the build file helped a lot in identifying and fixing them:
> 
> ```kotlin
> tasks.withType<Test> {
>     testLogging {
>         exceptionFormat = TestExceptionFormat.FULL
>         showStackTraces = true
>         showStandardStreams = true
>     }
> }
> ```
> 
> In our case the failing tests were mostly the ones that were using the Roboelectric runner.

I think this is the best way to go about it rather than using the work around of `runBlocking` because you can fix them once it for all and gives more confidence and deeper understanding to the team in having reliable test suites. In our case the room db instance wasn't closed in one of the test file's `After` block which was causing the issue so leaving the below code if it helps someone in the future.

`System.logW: A resource was acquired at attached stack trace but never released. See java.io.Closeable for information on avoiding resource leaks.
java.lang.Throwable: Explicit termination method 'close' not called`

`@After
    fun teardown() {
        inMemoryTestDbInstance.close()
    }`

**Comment by DickLemm** [comment]
We used runBlocking in this test (where suppresed exception was thrown), but other tests where using runTest and failed because of the supressed exception .

I solved it by doing in teardown:

@After
    fun tearDown() {
        Dispatchers.resetMain()
        testDispatcher.cleanupTestCoroutines()
        try {
            runTest {
                // runTest is checking for suppressed exceptions 
                // because we thrown an exception in this test, it would fail next test.
                // by running runTest here, we solve it.
            }
        } catch (e: Exception){}
    }

**Comment by mnmari** [comment]
Besides the issues with "no answer found for ...," I encountered additional errors related to missing mocks, such as "class java.lang.String cannot be cast to class kotlin.Result (java.lang.String is in module java.base of loader 'bootstrap'; kotlin.Result is in unnamed module of loader 'app')." I resolved these by mocking method returns with `coEvery` and setting `relaxed = true` when declaring my mocked classes.

After fixing all of these issues, I still faced some strange and generic errors with Coroutines in tests, which indicated that the main dispatcher was not set to a test dispatcher, even though I had done so.

Through troubleshooting, I discovered that calling Coroutines in an init block in some ViewModels was causing side effects in these tests. When I replaced the init block with a public method, these modules behaved more stably when running test suites.

Also, if you are replacing the main dispatcher with a test dispatcher, it’s important to ensure that `runTest` is used in all tests, even if they don’t contain explicit asynchronous calls, as noted [here](https://github.com/Kotlin/kotlinx.coroutines/issues/4141#issuecomment-2288306526).
